parser_library/src/context/using.cpp

```diff
@@ -0,0 +1,551 @@
+        ordinary_assembly_dependency_solver solver(ord_context, get(expr.context).evaluation_ctx);
+
+        expr.value = expr.expression->evaluate(solver);
+        if (expr.value.value_kind() == symbol_value_kind::UNDEF)
+            expr.label = identify_label(ord_context, expr.expression.get());
+
+        expr.expression->collect_diags();
+        if (expr.label == nullptr)
+        {
+            for (auto& d : expr.expression->diags())
+                diag.add_diagnostic(add_stack_details(std::move(d), get(expr.context).stack));
+        }
+        expr.expression->diags().clear();
+    }
+
+    for (auto& u : m_usings)
+    {
+        diagnostic_consumer_transform t([this, &diag, &u](diagnostic_op d) {
+            diag.add_diagnostic(add_stack_details(std::move(d), get(u.instruction_ctx).stack));
+        });
+        u.resolve(*this, t);
+        u.compute_context(*this, t);
+    }
+}
+
+index_t<using_collection::instruction_context> using_collection::add(
+    dependency_evaluation_context ctx, processing_stack_t stack)
+{
+    m_instruction_contexts.push_back({ std::move(ctx), std::move(stack) });
+    return index_t<instruction_context>(m_instruction_contexts.size() - 1);
+}
+
+index_t<using_collection::mach_expression> using_collection::add(
+    std::unique_ptr<const mach_expression> expr, index_t<instruction_context> ctx)
+{
+    m_expr_values.push_back({ std::move(expr), ctx });
+    return index_t<mach_expression>(m_expr_values.size() - 1);
+}
+
+
+index_t<using_collection> using_collection::add(index_t<using_collection> current,
+    id_index label,
+    std::unique_ptr<mach_expression> begin,
+    std::unique_ptr<mach_expression> end,
+    std::vector<std::unique_ptr<mach_expression>> args,
+    dependency_evaluation_context eval_ctx,
+    processing_stack_t stack)
+{
+    assert(!args.empty() && args.size() <= reg_set_size);
+
+    index_t<instruction_context> ctx_id = add(std::move(eval_ctx), std::move(stack));
+
+    auto b = add(std::move(begin), ctx_id);
+    auto e = end ? add(std::move(end), ctx_id) : index_t<mach_expression>();
+
+    std::vector<index_t<mach_expression>> base;
+    base.reserve(args.size());
+
+    std::transform(args.begin(), args.end(), std::back_inserter(base), [this, ctx_id](auto& a) {
+        return add(std::move(a), ctx_id);
+    });
+
+    m_usings.emplace_back(current, ctx_id, b, base, label, e);
+    return index_t<using_collection>(m_usings.size() - 1);
+}
+
+index_t<using_collection> using_collection::remove(index_t<using_collection> current,
+    std::vector<std::unique_ptr<mach_expression>> args,
+    dependency_evaluation_context eval_ctx,
+    processing_stack_t stack)
+{
+    index_t<instruction_context> ctx_id = add(std::move(eval_ctx), std::move(stack));
+
+    std::vector<index_t<mach_expression>> base;
+    base.reserve(args.size());
+
+    std::transform(args.begin(), args.end(), std::back_inserter(base), [this, ctx_id](auto& a) {
+        return add(std::move(a), ctx_id);
```

Comment:
I think something like `remove(...)` should be used here

Reply:
this is actually adding an expression to the vector of expressions, so I think add is appropriate.

Reply:
Ok